tests/generic_config_updater/test_aaa.py

```diff
@@ -0,0 +1,609 @@
+import logging
+import pytest
+
+from tests.common.helpers.assertions import pytest_assert
+from tests.generic_config_updater.gu_utils import apply_patch, expect_op_success, expect_op_failure
+from tests.generic_config_updater.gu_utils import generate_tmpfile, delete_tmpfile
+from tests.generic_config_updater.gu_utils import create_checkpoint, delete_checkpoint, rollback_or_reload
+
+pytestmark = [
+    pytest.mark.topology('t0'),
+]
+
+logger = logging.getLogger(__name__)
+
+AAA_CATEGORY = ["authentication", "authorization", "accounting"]
+
+@pytest.fixture(autouse=True)
+def setup_env(duthosts, rand_one_dut_hostname, tbinfo):
+    """
+    Setup/teardown fixture for each loopback interface test.
+    rollback to check if it goes back to starting config
+
+    Args:
+        duthosts: list of DUTs.
+        rand_selected_dut: The fixture returns a randomly selected DuT.
+    """
+    duthost = duthosts[rand_one_dut_hostname]
+
+    create_checkpoint(duthost)
+
+    yield
+
+    try:
+        logger.info("Rolled back to original checkpoint")
+        rollback_or_reload(duthost)
+    finally:
+        delete_checkpoint(duthost)
+
+
+
+def verify_aaa_sub_options(duthost, aaa_type, option, value):
+    """ Verify if AAA sub type's options match with expected value
+
+    Sample output:
+    admin@vlab-01:~/tacacs$ show aaa
+    AAA authentication login local (default)
+    AAA authentication failthrough False (default)
+    AAA authorization login local (default)
+    AAA accounting login disable (default)
+    """
+    output = duthost.shell('show aaa | grep -Po "AAA {} {} \K.*"'.format(aaa_type, option))
+    logger.info("{}".format(output))
+    pytest_assert(not output['rc'],
+        "Failed to grep AAA {}".format(option)
+    )
+    pytest_assert(output['stdout'] == value,
+        "AAA {} {} failed to apply".format(aaa_type, option)
+    )
+
+def aaa_add_init_config(duthost):
+    """ Add initial config for AAA
+
+    Though AAA has default value in setup. But the config does not
+    included in configDB. So to make change on AAA table, the init
+    config need to be added to config first.
```

Comment:
Do you mean generic config updater could not work if AAA table is missing in ConfigDB, ie, using the default values?

Reply:
GCU can work if AAA table is missing. But it have to be configured first in t0 to make that table appear in configDB.
If we want to REPLACE value of AAA table, we need to do the `replace` action which means the table need to exist and contains some value in advance.
The test case `test_aaa_tc1_add_config` which is to test adding init config thru GCU.
But for those tests with `replace` action, function `aaa_add_init_config(duthost)` is called to ensure we have the table and value to be replaced.
Please see below for configDB AAA table:
```
admin@vlab-01:~$ show aaa                                        
AAA authentication login local (default)
AAA authentication failthrough False (default)
AAA authorization login local (default)
AAA accounting login disable (default)
admin@vlab-01:~$ show run all | grep AAA -A3                      <=== default initial setup doesn't contain table in advance

admin@vlab-01:~$ sudo config aaa authentication login default     <=== though we manually configured it to default,
admin@vlab-01:~$ show run all | grep AAA -A3                      <=== but table still don't show up

admin@vlab-01:~$ sudo config aaa authentication login local       <=== Configured it to local(same as default value)
admin@vlab-01:~$ show run all | grep AAA -A3                      <=== Table show up so we can REPLACE the value by GCU
    "AAA": {
        "authentication": {
            "login": "local"
        }

admin@vlab-01:~$ sudo config aaa authentication login default     <=== set back to default
admin@vlab-01:~$ show run all | grep AAA -A3                      <===  table doesn't disappear but contians empty value
    "AAA": {
        "authentication": {}
    },
    "ACL_TABLE": {

```

Reply:
In real world, it is a quite common flow:
1. deploy the switch with basic config, and there is no AAA table
2. later we think the switch should connect to a TACACS server, so we need to config AAA, and TACACS config

Reply:
Sounds like a combination of `test_aaa_tc1_add_config(duthost, aaa_type, aaa_sub_options)` and `test_tacacs_global_tc5_add_config(duthost)` and `test_tacacs_server_tc9_add(duthost, ip_address)`, right?

Reply:
Added `add_init_config_XXX_without_table` to simulate that configDB doesn't have those tables.